### PR TITLE
Implement king

### PR DESCRIPTION
**Pull request by avani-k**
Final review before merging onto main.

### CODE REVIEW

File: king-local.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "king.gds_file": "gs://topmed_workflow_testing/UWGAC_WDL/ALL.pruned.gds"
```
Comment (aofarrel) [inline king-local.json:2]: Since this local JSON the exact same as the terra one, I don't think it's really needed. I only create local JSONs when I'm testing with big files I don't want to keep downloading from Google, so I just have them link to my filesystem instead.

File: testdata/README.md [hunk 2]
```diff
@@ -7,6 +7,8 @@ http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/
 
 The variants are in genome build hg19.
 
+Test data for king.wdl is about 149.9 Mb and available as ALL.pruned.gds in: gs://topmed_workflow_testing/UWGAC_WDL/ALL.pruned.gds
```
Comment (aofarrel) [inline testdata/README.md:10]: You may want to change this to the storage.google.cloud link in the README, as that is the form that internet browsers can resolve.

ie, https://storage.google.com/topmed_workflow_testing/UWGAC_WDL/ALL.pruned.gds

File: king-local.json (hunk 1 above)
Comment (avani-k) [reply king-local.json:2]: Do you suggest I just delete this particular json then?

File: testdata/README.md (hunk 2 above)
Comment (avani-k) [reply testdata/README.md]: Resolved in the last commit.

File: king/king.wdl [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+		preemptibles: "${preempt}"
+	}
+	output {
+		File king_matrix = glob("*.RData")[0]
+		File config_file = "king_to_matrix.config"
+	}
+
+}
+
+# [5] kinship_plots --
```
Comment (aofarrel) [inline king/king.wdl:278]: The name is fairly self-explanatory but you may want to put a short description of what this task does anyway.

File: king/king.wdl [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+	input {
+		File gds_file
+		File? sample_include_file
+		File? variant_include_file
+		String? out_prefix
+		File? phenotype_file
+		Float? kinship_plot_threshold
+		String? group
+		Float? sparse_threshold
+		String? kinship_method # not included in king-ibdseg-wf.cwl but is an input in king_to_matrix task
```
Comment (aofarrel) [inline king/king.wdl:377]: When this sort of thing happens, you actually don't have to define it as a workflow input. You can leave it as an input in the task's input section only and simply pretend it doesn't exist in the workflow section. On the Terra UI, it will still show up as a task-specific input. This will not only more closely mimic the cwl, but will also make it slightly clearer in Terra's UI.

**Review by aofarrel — CHANGES_REQUESTED** [review]
This is good work! My suggested changes are relatively small nitpicks.

File: king/checker/king-checker.wdl [hunk 5]
```diff
@@ -0,0 +1,113 @@
+version 1.0
+
+# Replace the first URL here with the URL of the workflow to be checked.
+import "https://raw.githubusercontent.com/DataBiosphere/analysis_pipeline_WDL/implement-king/king.wdl" as check_me
+import "https://raw.githubusercontent.com/aofarrel/checker-WDL-templates/v0.9.3/tasks/filecheck_task.wdl" as checker_file
+
+# There is no functional difference between "here's an array of files from
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/checker/king-checker.wdl:7]: Deleted!

File: king/checker/king-checker.wdl [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+# AND the array itself are NOT optional. You do not need to know how many
+# files are in an array, but none of those files can have type File?.
+
+workflow checker {
+	input {
+		# First set of inputs: The same input(s) as the workflow to be checked
+		File gds_file
+
+		# Second set of inputs: The truth file(s)
+		#File truth_processed_bed
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/checker/king-checker.wdl:19]: They are not actual outputs, they are outputs from the first two tasks that are used an input for the following tasks. Hence, we don't have truth files as such for these.

File: king/checker/king-checker.wdl [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		#File truth_bed_file
+		File truth_king_ibdseg_output
+		File truth_king_matrix
+		File truth_kinship_plots
+
+		# Optional inputs:
+		#File sample_include_file
+		#File variant_include_file
+		#String king_ibdseg_out_prefix = "king_ibdseg"
+		#Float sparse_threshold = 0.02209709
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/checker/king-checker.wdl:29]: Would you then recommend just removing them?

File: king/king-local.json [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
+  "king.gds_file": "gs://topmed_workflow_testing/UWGAC_WDL/ALL.pruned.gds"
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/king-local.json:2]: Would you recommend just removing this file entirely or leaving the space blank for someone trying to run it locally. Reason for this being that ALL.pruned.gds is too big to be hosted on GitHub. So I cannot really have any local path here.

File: king/checker/king-checker.wdl [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+		#String kinship_plots_kinship_method = "king"
+
+
+	}
+
+	# Task 1 - gds2bed
+	call check_me.gds2bed as gds2bed {
+		input:
+			gds_file = gds_file
+			#sample_include_file = sample_include_file,
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/checker/king-checker.wdl]: Deleted!

File: king/king.wdl [hunk 10]
```diff
@@ -0,0 +1,419 @@
+version 1.0
+
+# king.wdl -- This workflow uses the KING --ibdseg method to estimate kinship coefficients, and returns results for pairs of related samples. These kinship estimates can be used as measures of kinship in PC-AiR.
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/king.wdl:3]: Done!

File: king/checker/king-checker.wdl (hunk 7 above)
[the remark this answers is not in the document]
Comment (aofarrel) [reply king/checker/king-checker.wdl:19]: Yes, unless you want to run some tests involving non-default inputs.

It's a bit of an awkward situation as Stephanie's pipelines don't really have unit testing, but I think we can use checker workflows to do a little bit of unit testing for her without really giving us much more work -- all we have to do is put in a non-default value here and in the CWL and then check against the CWL's output as a new truth file. If there is a mismatch that can help us find bugs in the CWL or the WDL. For instance, I tested ten configurations of null model, and that's how we discovered that the R script is buggy when run on an AWS backend. Strictly speaking that isn't our problem, but it is a good thing to know and can be reported as a bug to them.

I would recommend running the whole workflow (or at least relevant steps) twice. The first time, don't pass in sparse_threshold which will make it act as if sparse_threshold is the default (0.02209709). The second time, pass in a new value to sparse_threshold. I don't know what some valid other options would be; ie, I don't know if something like 0.5 would be valid from a scientific point of view. I think sparse_threshold is supposed to represent some degree of relatedness, so you may want to do some searching to find out what number might represent more or less related individuals. Once you have a new number to test, run the CWL with that number to make a new set of truth files and just add another test to your checker.

File: king/king.wdl [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+		set -eux -o pipefail
+
+		echo "Generating bash executable file"
+		python << CODE
+		import os
+
+		f = open("plink_cmd.sh", "a")
+
+		f.write('#!/bin/bash\n')
+		f.write('plink --make-bed --bfile ' + '.'.join("~{bimfile}".split('.')[:-1]) + ' --bed ' + "~{bedfile}" + ' --fam ' + "~{famfile}" + ' --out ' + os.path.basename('.'.join(("~{bedfile}").split('.')[:-1])) + '_recode\n')
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/king.wdl:110]: Done using \

File: king/king.wdl [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+		python << CODE
+		import os
+
+		# king --ibdseg -b "~{bed_file}" --cpus "~{cpu}" --prefix "~{out_prefix}"
+		g = open("king_ibdseg_cmd.sh", "a")
+
+		g.write('#!/bin/bash\n')
+
+		if "~{out_prefix}" != "":
+			g.write('king --ibdseg -b ' + "~{bed_file}" + ' --bim ' + "~{bim_file}" + ' --fam ' + "~{fam_file}"  + ' --cpus ' + "~{cpu}" + ' --prefix ' + "~{out_prefix}" + '\n')
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/king.wdl:169]: Done using \

File: king/king.wdl [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+	}
+	output {
+		File king_ibdseg_output = glob("*.seg")[0]
+	}
+
+}
+
+# [4] king_to_matrix -- 
+
+# Add java script from cwl?
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/king.wdl:195]: Yes, remember the weird metadata javascript that doesnt seem to be doing anything? I thought I should leave some reminder that it exists. If we want to keep it, I can rephrase it to make it more clear. Or else, get rid of that comment.

File: king/king.wdl (hunk 3 above)
Comment (avani-k) [reply king/king.wdl:278]: Done!

File: king/king.wdl [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+			sparse_threshold = sparse_threshold,
+			out_prefix = out_prefix,
+			kinship_method = kinship_method
+	}	
+	call kinship_plots{
+		input:
+			kinship_file = king_ibdseg.king_ibdseg_output,
+			phenotype_file = phenotype_file,
+			sample_include_file = sample_include_file,
+			kinship_method = kinship_method
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/king.wdl:411]: Gotcha!

File: king/king.wdl (hunk 4 above)
Comment (avani-k) [reply king/king.wdl:377]: Makes sense, done!!

File: king/checker/king-checker.wdl [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+		# Second set of inputs: The truth file(s)
+		#File truth_processed_bed
+		#File truth_bed_file
+		File truth_king_ibdseg_output
+		File truth_king_matrix
+		File truth_kinship_plots
+
+		# Optional inputs:
+		#File sample_include_file
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/checker/king-checker.wdl]: Done!

File: king/king.wdl [hunk 16]
```diff
@@ -0,0 +1,419 @@
...
+		File processed_bim = glob("*.bim")[0]
+		File processed_fam = glob("*.fam")[0]
+		File config_file = "gds2bed.config"
+	}
+
+}
+
+# [2] plink_make_bed -- process bed file through plink
+
+# Add java script from cwl?
```
[the remark this answers is not in the document]
Comment (avani-k) [reply king/king.wdl:80]: See comment below.

File: king/king-local.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (aofarrel) [reply king/king-local.json:2]: I'd recommend deleting king-local.json entirely.

File: king/king.wdl (hunk 13 above)
[the remark this answers is not in the document]
Comment (aofarrel) [reply king/king.wdl:204]: I think it's best change the comment to something like "this task skips the seven bridges-specific inheritance of metadata," both here and around line 80.

File: king/king.wdl [hunk 17]
```diff
@@ -0,0 +1,425 @@
...
+		
+		# runtime attributes
+		Int addldisk = 5
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+
+	# Estimate disk size required
+	Int gds_size = ceil(size(gds_file, "GB"))
```
Comment (aofarrel) [inline king/king.wdl:27]: It is unlikely for users to have a large sample_include_file or a large variant_include_file, but is technically possible. Add some calculations for those files. Since they're optional, you can use select_first(). 

Let me just save you some time here: combining select_first() with size() and ceil() is kind of a pain, and it seems this is the only way to do it:

Int smpl_size = select_first([ceil(size(sample_include_file, "GB")), 0])

ie, select_first, then ceil, then size. Try to do it in any other order and it will probably break. In this scenario, if sample_include_file is not defined, then in theory smple_size will be 0.

In addition to accounting for the sample include and the variant include, don't forget about your output file. Do you expect your bim, bam, and fam to total less than the size of your input GDS? If so, this is easy, just double the size of your GDS input in your calculation of final_disk_size. This will be trickier to estimate if you expect your outputs to be bigger than your input... try to gauge a good ratio from your test runs and based on that improve the formula for final_disk_size. It doesn't have to be perfect, especially as users can slap on a few extra gigs with addldisk, but as-written this is likely to fail on "real" data which is larger than what we're testing with.

File: king/king.wdl [hunk 18]
```diff
@@ -0,0 +1,425 @@
...
+		
+		# runtime attributes
+		Int addldisk = 5
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+
+	# Estimate disk size required
+	Int bed_size = ceil(size(bedfile, "GB"))
```
Comment (aofarrel) [inline king/king.wdl:99]: Since you're ending up with a reencoded version of your bed input here (I think? I don't know a lot about plink) you should have bed_size get at least doubled in your final_disk_size calculation. It is a good idea to also account for the size of the bim and fam (although I believe they tend to be less than a gig).

File: king/king.wdl [hunk 19]
```diff
@@ -0,0 +1,425 @@
...
+		
+		# runtime attributes
+		Int addldisk = 5
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+
+	# Estimate disk size required
+	Int bed_size = ceil(size(bed_file, "GB"))
```
Comment (aofarrel) [inline king/king.wdl:158]: Based on the truth file you uploaded, I think there's a decent chance the seg file will be pretty big when run on "real" data. As with gds2bed, try to gauge the size of the output file based on the size of the input bed and account for that in your final_disk_size calc.

File: king/king.wdl [hunk 20]
```diff
@@ -0,0 +1,425 @@
...
+		Int addldisk = 5
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+
+	# Estimate disk size required
+	Int king_size = ceil(size(king_file, "GB"))
+	Int final_disk_dize = king_size + addldisk
+
```
Comment (aofarrel) [inline king/king.wdl:226]: See earlier comments r/e disk size calcs.

File: king/king.wdl [hunk 21]
```diff
@@ -0,0 +1,425 @@
...
+		
+		# runtime attributes
+		Int addldisk = 5
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+
+	# Estimate disk size required
+	Int kinship_size = ceil(size(kinship_file, "GB"))
```
Comment (aofarrel) [inline king/king.wdl:309]: I think your outputs are always going to be relatively small, but still consider accounting for your optional inputs. phenotype_file in particular can be quite large.

File: king/checker/king-checker.wdl [hunk 22]
```diff
@@ -0,0 +1,87 @@
+version 1.0
+
+# Replace the first URL here with the URL of the workflow to be checked.
```
Comment (aofarrel) [inline king/checker/king-checker.wdl:3]: Delete this comment, it might confuse users.

**Review by aofarrel — CHANGES_REQUESTED** [review]
Thanks for making the changes so quickly. A few improvements to the disk size calculations are in order.

File: king/checker/king-checker.wdl (hunk 7 above)
[the remark this answers is not in the document]
Comment (avani-k) [reply king/checker/king-checker.wdl:19]: I will remove them for now. This can be a task to take up in the future.
